### PR TITLE
New semantic analyzer: fix issues with 'X = X' assignments

### LINKED ISSUE #6404: New semantic analyzer: Support idem forward aliases
This is a follow-up for #6390

This tricky corner case is used in some real code and is supported by old semantic analyzer
```python
C = C
class C:  # type: ignore
    pass

x: C
```
But with new analyzer this results in `Invalid type` since the first definition always wins. Probably we just need to special-case this somehow.

### PR DESCRIPTION
**Pull request by JukkaL**
We special case these since a more general fix, such as looking
at line numbers of definitions, would break a ton of test cases.

Fixes #6404.

### CODE REVIEW

File: mypy/newsemanal/semanal.py [hunk 1]
```diff
@@ -1868,6 +1873,46 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
...
+            return False
+        # It's an assignment like 'X = X' in the global scope.
+        name = lvalue.name
+        sym = self.lookup(name, s)
+        if sym is None:
+            if self.final_iteration:
+                # Fall back to normal assignment analysis.
+                return False
+            else:
+                self.defer()
```
Comment (ilevkivskyi) [inline mypy/newsemanal/semanal.py]: `name_not_defined()` currently already calls `defer()`, is this additional defer actually needed?

Also this returns `True`, I would make it more clear by putting the return directly here.

File: mypy/newsemanal/semanal.py [hunk 2]
```diff
@@ -1868,6 +1873,46 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
...
+                return False
+            if name not in self.globals:
+                # The name is from builtins. Add an alias to the current module.
+                self.add_symbol(name, sym.node, s)
+            if not isinstance(sym.node, PlaceholderNode):
+                for node in s.rvalue, lvalue:
+                    node.node = sym.node
+                    node.kind = GDEF
+                    node.fullname = sym.node.fullname()
+        return True
```
Comment (ilevkivskyi) [inline mypy/newsemanal/semanal.py]: With the above this return can be indented one level more.

File: test-data/unit/check-newsemanal.test [hunk 3]
```diff
@@ -2886,3 +2886,85 @@ def f() -> None:
...
+[case testNewAnalyzerIdentityAssignment7]
+C = C # E: Name 'C' is not defined
+
+reveal_type(C) # N: Revealed type is 'Any' \
+               # E: Name 'C' is not defined
+
+[case testNewAnalyzerIdentityAssignment8]
+from typing import Final
+x: Final = 0
+x = x # E: Cannot assign to final name "x"
```
Comment (ilevkivskyi) [inline test-data/unit/check-newsemanal.test]: Great to see so detailed test cases!

File: mypy/newsemanal/semanal.py (hunk 1 above)
Comment (JukkaL) [reply mypy/newsemanal/semanal.py]: The fact that `name_not_defined` calls defer requires some pretty subtle reasoning so I'd rather keep make it explicit here.

**Pull request merged**